### ISSUE TITLE
Required site properties are None after packaging an executable with PyInstaller

### ISSUE THREAD

**Issue by heidi-jack** [pyinstaller/pyinstaller#8878]
I have a simple Kivy application, with some added KivyMD components. Here's my pip freeze output...

```
altgraph==0.17.4
certifi==2024.8.30
charset-normalizer==3.4.0
configparser==7.1.0
Cython==3.0.0
docutils==0.21.2
idna==3.10
Kivy==2.3.0
kivy-deps.angle==0.4.0
kivy-deps.glew==0.3.1
kivy-deps.sdl2==0.7.0
Kivy-examples==2.3.0
Kivy-Garden==0.1.5
kivy-translate==0.2.0
kivymd==1.2.0
macholib==1.16.3
modulegraph==0.19.6
packaging==24.2
pefile==2023.2.7
pillow==10.4.0
py2app==0.28.8
Pygments==2.18.0
pyinstaller==6.11.1
pyinstaller-hooks-contrib==2024.10
pypiwin32==223
pywin32==308
pywin32-ctypes==0.2.3
requests==2.32.3
urllib3==2.2.3
```

I have followed the instructions for packaging on Windows found [here](https://kivy.org/doc/stable/guide/packaging-windows.html). I have already got a version working with MacOS, so using what I learned from that I tried to package for Windows.

Here's my main.spec file, which as per the Kivy docs, I run from a folder within my project folder, hence the ..\main.py usages:

```
# -*- mode: python ; coding: utf-8 -*-

from kivy_deps import sdl2, glew
from kivy import kivy_data_dir
from kivymd import hooks_path as kivymd_hooks_path
block_cipher = None

# assets
kivy_assets_toc = Tree(kivy_data_dir, prefix=join('kivy_install', 'data'))
tocs = [kivy_assets_toc]

a = Analysis(
    ['..\\main.py'],
    pathex=[],
    binaries=[],
    datas=[('..\\locale\\*', 'locale'),('..\\images\\*', 'images')],
    hiddenimports=[],
    hookspath=[kivymd_hooks_path],
    hooksconfig={},
    runtime_hooks=["hook.py"],
    excludes=['_tkinter', 'Tkinter', 'enchant', 'twisted'],
    noarchive=False,
    cipher=block_cipher,
    optimize=0,
)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

exe = EXE(
    pyz,
    Tree('..\\'),
    a.scripts,
    *[Tree(p) for p in (sdl2.dep_bins + glew.dep_bins)],
    *tocs,
    exclude_binaries=True,
    name='main',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    *tocs,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='main',
)
```

Which also has a hook.py that sets an environment variable:

```
import os
from os.path import join

# Required to fix OpenGL version bug when building in Windows
# where when running and freezing the application, the application cannot find the correct
# OpenGL version and fails. This forces it to work for the build...
os.environ['KIVY_GL_BACKEND'] = 'angle_sdl2'
```

Now, once packaged and run, I get this error from angle and a similar one for sdl2:

```
Traceback (most recent call last):
  File "main.py", line 21, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 384, in exec_module
  File "kivymd\__init__.py", line 26, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 384, in exec_module
  File "kivy\__init__.py", line 320, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 384, in exec_module
  File "kivy_deps\angle\__init__.py", line 23, in <module>
  File "<frozen ntpath>", line 108, in join
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

This comes from the angle.__init__ and sdl2.__init__ code which looks like below, and uses site.USER_BASE and site.USER_SITE respectively. It fails on the join due to d being None when looking for site.USER_BASE

```
for d in [sys.prefix, site.USER_BASE]:
    p = join(d, 'share', 'sdl2', 'bin')
    if isdir(p):
        os.environ["PATH"] = p + os.pathsep + os.environ["PATH"]
        if hasattr(os, 'add_dll_directory'):
            os.add_dll_directory(p)
        dep_bins.append(p)
```

It turns out that site.USER_BASE and site.USER_SITE are None when packaged. From what I am reading is that pyinstaller mocks out the site package: https://github.com/pyinstaller/pyinstaller/issues/510

Maybe my knowledge of python isn't great, but surely we need these set within a python application for packages to find python base and the site-packages?

I can optionally overwrite the values in my main.py file before the rest of the packages are imported, but not sure what they should be set to when within an executable? Note that when I add the following the executable builds, but running it fails with no error message. It just seems to crash. Understandably as i've set these to an empty string.

```
import site
if site.USER_SITE is None:
    site.USER_SITE = ""

if site.USER_BASE is None:
    site.USER_BASE = ""
```

Is this impossible given that PyInstaller mocks the site import? Anyone else been successful with this?

**Comment by rokm** [comment]
> Maybe my knowledge of python isn't great, but surely we need these set within a python application for packages to find python base and the site-packages?

PyInstaller's embedded python interpreter runs in the isolated mode, so with disabled site-packages. Which is why `site.USER_BASE` and `site.USER_SITE` are not initialized.

We do not use mock `site` module anymore, but rather the stdlib one. So if a package fails to account for possibility of those variables being `None`, that's a bug on their side.

> I can optionally overwrite the values in my main.py file before the rest of the packages are imported, but not sure what they should be set to when within an executable? Note that when I add the following the executable builds, but running it fails with no error message. It just seems to crash. Understandably as i've set these to an empty string.

Is the crash really related the those `site` variables? It could as well be that you got past the `None` issue and are now facing something else. 

For what is worth, you could try setting those paths to `sys._MEIPASS` in the frozen application (or even `sys.prefix`).

**Comment by heidi-jack** [comment]
The crash once I set them to `""` is definitely due to those being an empty string, but just wanted to prove out that those being None is definitely the cause of the previous crash.

Thanks for your advice, i'll have a go and try my best. It seems odd that it would be a bug in the packages themselves as Kivy is a popular UI tool and so is PyInstaller, so I assumed it's something i've done wrong. I'll try your suggestions and report back.

**Comment by rokm** [comment]
But actually, your spec looks pretty shady, too...

You seem to be building `onedir` application (based on `COLLECT`), but you are passing some of the data tocs to `EXE` as well. So you are likely getting an application with `onefile` semantics that ends up missing  elements that are in `a.datas` and `a.binaries`.

**Comment by heidi-jack** [comment]
Ha yes, it's been 3 hours of trying everything, onefile, not onefile, add this line in, add this other line in. The spec file has become unwieldy and i've lost track at this point of what i've added and what's needed. Maybe I should start from scratch. 

If it helps, doing a build with just the following also has the same issue:


```powershell
$env:KIVY_GL_BACKEND="angle_sdl2"; pyinstaller --clean --noconfirm --windowed --hidden-import=kivy_deps.sdl2 --hidden-import=kivy_deps.glew --hidden-import=kivy_deps.angle --hidden-import=kivy_deps.gstreamer --name main ..\main.py
```

**Comment by rokm** [comment]
FWIW, I can freeze and run the following example:

```python
import sys
if getattr(sys, 'frozen', False):
    import site
    site.USER_BASE = ''
    site.USER_SITE = ''

from kivy.app import App
from kivy.uix.widget import Widget

class PongGame(Widget):
    pass

class PongApp(App):
    def build(self):
        return PongGame()

if __name__ == '__main__':
    PongApp().run()
```

with the following spec:

```
# -*- mode: python ; coding: utf-8 -*-

from kivy_deps import sdl2, glew, angle

a = Analysis(
    ['program.py'],
    pathex=[],
    binaries=[],
    datas=[],
    hiddenimports=['kivy_deps.angle', 'kivy_deps.glew', 'kivy_deps.sdl2'],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False,
    optimize=0,
)
pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='program',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    *[Tree(p) for p in (sdl2.dep_bins + glew.dep_bins + angle.dep_bins)],
    strip=False,
    upx=True,
    upx_exclude=[],
    name='program',
)
```

The `None` work-around for `user.USER_BASE` / `user.USER_SITE` became necessary only after I added `kivy_deps.angle` /  `kivy_deps.sdl2` /  `kivy_deps.glew` to hidden imports.

But it actually also worked with just `pyinstaller --clean --noconfirm program.py` - presumably because contemporary PyInstaller versions are able to find SDL2/GLEW DLLS in the environment's shared directory without Kivy helpers. (But this build does not automatically pick up Angle DLLs).

**Comment by heidi-jack** [comment]
Thank you for your advice, you are right. Confirmed that the user.USER_BASE / user.USER_SITE workaround allows the application to continue, but now it's the application itself that's failing. 

 I managed to confirm by building a console application (my logger was borked too it seems...) but we are successfully past the user.USER_BASE / user.USER_SITE hurdle. I've started from scratch, and i've undone some of the complexity I added along the way that didn't fix the initial problem. 

I'm hoping the "empty" values of USER_BASE and USER_SITE aren't causing the new issue, but either way this is nothing to do with pyinstaller. A new error is a welcomed error.

I'll come back and post my final `.spec` file and work around for anyone else experiencing this.
